packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,303 @@
+import path from "node:path";
+import fg from "fast-glob";
+
+import type { ConfigRoute, DefineRouteFunction, RouteManifest } from "./routes";
+import { createRouteId, defineRoutes } from "./routes";
+import {
+  escapeEnd,
+  escapeStart,
+  isSegmentSeparator,
+  optionalEnd,
+  optionalStart,
+  paramPrefixChar,
+  routeModuleExts,
+} from "./routesConvention";
+
+export function flatRoutes(
+  appDirectory: string,
+  ignoredFilePatterns?: string[]
+): RouteManifest {
+  let extensions = routeModuleExts.join(",");
+
+  let routePaths = fg.sync(`*{${extensions}}`, {
+    absolute: true,
+    cwd: path.join(appDirectory, "routes"),
+    ignore: ignoredFilePatterns,
+  });
+
+  return flatRoutesUniversal(appDirectory, routePaths);
+}
+
+interface RouteInfo extends ConfigRoute {
+  name: string;
+  segments: string[];
+}
+
+/**
+ * Create route configs from a list of routes using the flat routes conventions.
+ * @param {string} appDirectory - The absolute root directory the routes were looked up from.
+ * @param {string[]} routePaths - The absolute route paths.
+ * @param {string} [prefix=routes] - The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+): RouteManifest {
+  let routeMap = getRouteMap(appDirectory, routePaths, prefix);
+
+  let uniqueRoutes = new Map<string, string>();
+
+  function defineNestedRoutes(
+    defineRoute: DefineRouteFunction,
+    parentId?: string
+  ): void {
+    let childRoutes = Array.from(routeMap.values()).filter(
+      (routeInfo) => routeInfo.parentId === parentId
+    );
+    let parentRoute = parentId ? routeMap.get(parentId) : undefined;
+    let parentRoutePath = parentRoute?.path ?? "/";
+    for (let childRoute of childRoutes) {
+      let routePath = childRoute.path?.slice(parentRoutePath.length) ?? "";
+      // remove leading slash
+      routePath = routePath.replace(/^\//, "");
+      let index = childRoute.index;
+      let fullPath = childRoute.path;
+
+      // add option to check for unique route ids
+      // this is copied from remix default convention
+      // but it is currently breaking some flat routes
+      // so until we can figure out a better way to do this
+      // make it optional to unblock users
+      let uniqueRouteId = (fullPath || "") + (index ? "?index" : "");
+      if (uniqueRouteId) {
+        let conflict = uniqueRoutes.get(uniqueRouteId);
+        if (conflict) {
+          throw new Error(
+            `Path ${JSON.stringify(fullPath)} defined by route ${JSON.stringify(
+              childRoute.id
+            )} conflicts with route ${JSON.stringify(conflict)}`
+          );
+        } else {
+          uniqueRoutes.set(uniqueRouteId, childRoute.id);
+        }
```

Comment:
nit: this block can be unnested since the `if` branch throws:

```suggestion
        uniqueRoutes.set(uniqueRouteId, childRoute.id);
```